### DISCUSSION TITLE
Tried to run our first c2m2 submission utilizing the cfde-submit cli, getting a 404 GlobusAPIError

### DISCUSSION THREAD

**Discussion by ACharbonneau** [nih-cfde/published-documentation#184] [Q&A] [answered]
Error message:

```
(cfde_venv) heathap@909c4ac4d83a c2m2 % cfde-submit login
You are already logged in
(cfde_venv) heathap@909c4ac4d83a c2m2 % cfde-submit run c2m2_submission/
Please enter the CFDE identifier for your Data Coordinating Center: cfde_registry_dcc:kidsfirst
Thank you. Would you like to save cfde_registry_dcc:kidsfirst for future submissions? (y/n)? > y
(404, 'Error', '{"error":{"code":"NOT_FOUND","detail":"Not found: /c636de1b-c7a8-4a93-bdec-ab4720328564"}}\n')
Traceback (most recent call last):
  File "/Users/heathap/projects/c2m2/cfde_venv/lib/python3.8/site-packages/cfde_submit/main.py", line 96, in run
    login_user()
  File "/Users/heathap/projects/c2m2/cfde_venv/lib/python3.8/site-packages/cfde_submit/main.py", line 188, in login_user
    cfde.check()
  File "/Users/heathap/projects/c2m2/cfde_venv/lib/python3.8/site-packages/cfde_submit/client.py", line 221, in check
    self.flow_client.get_flow(flow_info["flow_id"])
  File "/Users/heathap/projects/c2m2/cfde_venv/lib/python3.8/site-packages/globus_automate_client/flows_client.py", line 317, in get_flow
    return self.get(path, **kwargs)
  File "/Users/heathap/projects/c2m2/cfde_venv/lib/python3.8/site-packages/globus_sdk/base.py", line 215, in get
    return self._request(
  File "/Users/heathap/projects/c2m2/cfde_venv/lib/python3.8/site-packages/globus_sdk/base.py", line 502, in _request
    raise self.error_class(r)
globus_sdk.exc.GlobusAPIError: (404, 'Error', '{"error":{"code":"NOT_FOUND","detail":"Not found: /c636de1b-c7a8-4a93-bdec-ab4720328564"}}\n')
Error while starting Flow: GlobusAPIError(404, 'Error', '{"error":{"code":"NOT_FOUND","detail":"Not found: /c636de1b-c7a8-4a93-bdec-ab4720328564"}}\n')
```

**Comment by ACharbonneau** [comment, marked as answer]
This 404 error is due to incorrect permissions. The user is trying to do a submission, but does not have "Submitter" credentials.

Possible Solutions:

1. Use our [Onboarding to the Submission System instructions](https://github.com/nih-cfde/published-documentation/wiki/Onboarding-to-the-CFDE-Portal-Submission-System) to change your permissions. Only users with the Submitter role can push data to the submission system
2. Accept your invitation. If you have already sent in a request for Submitter status, but are getting this error, be sure that you fully accepted the Globus invitation to your Submitter group. You will need to click the 'Click here to apply for membership' text in the invitation message and follow instructions there before doing a submission.

**Reply by jeremyjyang** [reply]
I have a similar or same error. But I understood that I was an approved submitter. Could be related to how Globus supports linked accounts, so I am known via Orcid and Google.

+ cfde-submit login
You are already logged in
+ cfde-submit run /home/jjyang/src/idg-cfde/data/targetpages684/submission --dry-run --dcc-id cfde_registry_dcc:idg --output-dir /home/jjyang/src/idg-cfde/data/output --verbose
[DEBUG] cfde_submit.main::run() No previous state found
[DEBUG] cfde_submit.main::run() Determining DCC
[DEBUG] cfde_submit.main::run() Initializing Flow
[DEBUG] cfde_submit.client::check() (404, 'Error', '{"error":{"code":"NOT_FOUND","detail":"Not found: /c636de1b-c7a8-4a93-bdec-ab4720328564"}}\n')
[INFO] cfde_submit.client::check() Check FAILED, client lacks permissions or is not logged in.
[ERROR] cfde_submit.main::run() (404, 'Error', '{"error":{"code":"NOT_FOUND","detail":"Not found: /c636de1b-c7a8-4a93-bdec-ab4720328564"}}\n')
Traceback (most recent call last):
  File "/home/jjyang/.local/lib/python3.8/site-packages/cfde_submit/main.py", line 96, in run
    login_user()
  File "/home/jjyang/.local/lib/python3.8/site-packages/cfde_submit/main.py", line 188, in login_user
    cfde.check()
  File "/home/jjyang/.local/lib/python3.8/site-packages/cfde_submit/client.py", line 221, in check
    self.flow_client.get_flow(flow_info["flow_id"])
  File "/home/jjyang/.local/lib/python3.8/site-packages/globus_automate_client/flows_client.py", line 317, in get_flow
    return self.get(path, **kwargs)
  File "/home/jjyang/.local/lib/python3.8/site-packages/globus_sdk/base.py", line 215, in get
    return self._request(
  File "/home/jjyang/.local/lib/python3.8/site-packages/globus_sdk/base.py", line 502, in _request
    raise self.error_class(r)
globus_sdk.exc.GlobusAPIError: (404, 'Error', '{"error":{"code":"NOT_FOUND","detail":"Not found: /c636de1b-c7a8-4a93-bdec-ab4720328564"}}\n')
Error while starting Flow: GlobusAPIError(404, 'Error', '{"error":{"code":"NOT_FOUND","detail":"Not found: /c636de1b-c7a8-4a93-bdec-ab4720328564"}}\n')

**Reply by ACharbonneau** [reply]
I just checked and you are listed as a submitter for IDG, but it does look like you might have multiple Globus identities. Maybe we added the wrong one. Could you try:

- Go to https://app.nih-cfde.org/ and logging in with the same credentials you use in the tool (login is in the top right corner)
- Once logged in, click your name (top right corner)
- Select "My Profile"
- Check to see that you have the group "NIH CFDE IDG Submitters" listed

**Reply by jeremyjyang** [reply]
Following your instructions, I see I am in the submitters group as
0000-0002-1476-6192@orcid.org
The reason I tried another identity with Globus is that the ORCID identity resulted in a different error, apparently a Globus+ORCID error (screenshot attached). I got this error several times, from multiple computers, before trying my Google Globus identity.
![Screenshot from 2021-03-22 Globus ERROR](https://user-images.githubusercontent.com/864556/112063265-bd8bb580-8b26-11eb-898e-3418addbf788.png)

--

**Reply by jeremyjyang** [reply]
Progress. Seems the Globus identity issue resolved -- thanks @ACharbonneau, now new error. Is this because I'm missing the 22 blank header files? I'll try that next.

+ cfde-submit login
You are already logged in
+ cfde-submit run /home/jjyang/src/idg-cfde/data/targetpages684/submission --dry-run --dcc-id cfde_registry_dcc:idg --output-dir /home/jjyang/src/idg-cfde/data/output --verbose
[DEBUG] cfde_submit.main::run() No previous state found
[DEBUG] cfde_submit.main::run() Determining DCC
[DEBUG] cfde_submit.main::run() Initializing Flow
[INFO] cfde_submit.client::check() Check PASSED, client is ready use flows.
[DEBUG] cfde_submit.main::run() CfdeClient initialized, starting Flow
Submit datapackage 'submission' using cfde_registry_dcc:idg? (y/n)? > y
[INFO] cfde_submit.client::check() Check PASSED, client is ready use flows.
[DEBUG] cfde_submit.client::start_deriva_flow() Startup: Validating input
[DEBUG] cfde_submit.bdbag_utils::get_bag() Checking for a Git repository
[DEBUG] cfde_submit.bdbag_utils::get_bag() Git repo found, collecting metadata
[DEBUG] cfde_submit.bdbag_utils::get_bag() Creating BDBag out of directory '/home/jjyang/src/idg-cfde/data/targetpages684/submission'
[DEBUG] cfde_submit.bdbag_utils::get_bag() Copying data to '/home/jjyang/src/idg-cfde/data/targetpages684/submission_5d21b78d9219cafa660c0384894a5d639fa44dea' before creating BDBag
[DEBUG] cfde_submit.bdbag_utils::get_bag() BDBag created at '/home/jjyang/src/idg-cfde/data/targetpages684/submission_5d21b78d9219cafa660c0384894a5d639fa44dea'
[DEBUG] cfde_submit.bdbag_utils::get_bag() Archiving BDBag at '/home/jjyang/src/idg-cfde/data/targetpages684/submission_5d21b78d9219cafa660c0384894a5d639fa44dea' using 'zip'
[DEBUG] cfde_submit.bdbag_utils::get_bag() BDBag archived to file '/home/jjyang/src/idg-cfde/data/targetpages684/submission_5d21b78d9219cafa660c0384894a5d639fa44dea.zip'
[DEBUG] cfde_submit.bdbag_utils::get_bag() Removing old directory '/home/jjyang/src/idg-cfde/data/targetpages684/submission_5d21b78d9219cafa660c0384894a5d639fa44dea'
[DEBUG] cfde_submit.validation::validate_user_submission() Validating TableSchema in BDBag '/home/jjyang/src/idg-cfde/data/targetpages684/submission_5d21b78d9219cafa660c0384894a5d639fa44dea.zip'
TableSchema invalid due to the following errors:

**Reply by ACharbonneau** [reply]
Yes. If you don't have the blank tables, that's likely causing it. We're having some problems with certain errors not displaying in the tool. I'll add this one to the list. If you hit a wall and it's not giving you a useful error and you aren't sure why, send me a zip file of the package and I can run some tests against it to find the error.

**Reply by ACharbonneau** [reply]
> This 404 error is due to incorrect permissions. The user is trying to do a submission, but does not have "Submitter" credentials.
> 
> Possible Solutions:
> 
> 1. Use our [Onboarding to the Submission System instructions](https://github.com/nih-cfde/published-documentation/wiki/Onboarding-to-the-CFDE-Portal-Submission-System) to change your permissions. Only users with the Submitter role can push data to the submission system
> 2. Accept your invitation. If you have already sent in a request for Submitter status, but are getting this error, be sure that you fully accepted the Globus invitation to your Submitter group. You will need to click the 'Click here to apply for membership' text in the invitation message and follow instructions there before doing a submission.

The 404 error text has been updated as of 3/30/2021 to read:
![image](https://user-images.githubusercontent.com/1719360/114621461-b7e34480-9c7a-11eb-9c58-9ed80a0ff953.png)

**Comment by jeremyjyang** [comment]
Hi Amanda,

The Globus error seems solved, but still there is another error.

Here's a zip of the datapackage:

https://www.dropbox.com/s/8m09rtdh1ohnikk/idg_submission.zip?dl=0

All 22 files are there, with data only in file.tsv

> (cfde) lengua$ ls -l $DATAPATH/
> total 3904
> -rw-rw-r-- 1 jjyang jjyang      20 Mar 23 12:27 anatomy.tsv
> -rw-rw-r-- 1 jjyang jjyang      20 Mar 23 12:27 assay_type.tsv
> -rw-rw-r-- 1 jjyang jjyang      80 Mar 23 12:27 biosample_from_subject.tsv
> -rw-rw-r-- 1 jjyang jjyang      86 Mar 23 12:27 
> biosample_in_collection.tsv
> -rw-rw-r-- 1 jjyang jjyang      96 Mar 23 12:27 biosample.tsv
> -rw-rw-r-- 1 jjyang jjyang   52464 Mar 22 13:49 C2M2_datapackage.json
> -rw-rw-r-- 1 jjyang jjyang      82 Mar 23 12:27 
> collection_defined_by_project.tsv
> -rw-rw-r-- 1 jjyang jjyang     120 Mar 23 12:27 
> collection_in_collection.tsv
> -rw-rw-r-- 1 jjyang jjyang      80 Mar 23 12:27 collection.tsv
> -rw-rw-r-- 1 jjyang jjyang      20 Mar 23 12:27 data_type.tsv
> -rw-rw-r-- 1 jjyang jjyang      74 Mar 23 12:27 
> file_describes_biosample.tsv
> -rw-rw-r-- 1 jjyang jjyang      70 Mar 23 12:27 file_describes_subject.tsv
> -rw-rw-r-- 1 jjyang jjyang      20 Mar 23 12:27 file_format.tsv
> -rw-rw-r-- 1 jjyang jjyang      76 Mar 23 12:27 file_in_collection.tsv
> -rw-rw-r-- 1 jjyang jjyang 3857938 Mar 23 12:27 file.tsv
> -rw-rw-r-- 1 jjyang jjyang      33 Mar 23 12:27 id_namespace.tsv
> -rw-rw-r-- 1 jjyang jjyang      26 Mar 23 12:27 ncbi_taxonomy.tsv
> -rw-rw-r-- 1 jjyang jjyang     115 Mar 23 12:27 primary_dcc_contact.tsv
> -rw-rw-r-- 1 jjyang jjyang     102 Mar 23 12:27 project_in_project.tsv
> -rw-rw-r-- 1 jjyang jjyang      80 Mar 23 12:27 project.tsv
> -rw-rw-r-- 1 jjyang jjyang      82 Mar 23 12:27 subject_in_collection.tsv
> -rw-rw-r-- 1 jjyang jjyang      58 Mar 23 12:27 subject_role_taxonomy.tsv
> -rw-rw-r-- 1 jjyang jjyang     100 Mar 23 12:27 subject.tsv

 From today's latest try:

> (cfde) lengua$ cfde-submit run $DATAPATH --dry-run --dcc-id 
> cfde_registry_dcc:idg --output-dir $DATADIR/output --verbose
> [DEBUG] cfde_submit.main::run() No previous state found
> [DEBUG] cfde_submit.main::run() Determining DCC
> [DEBUG] cfde_submit.main::run() Initializing Flow
> [INFO] cfde_submit.client::check() Check PASSED, client is ready use 
> flows.
> [DEBUG] cfde_submit.main::run() CfdeClient initialized, starting Flow
> Submit datapackage 'submission' using cfde_registry_dcc:idg? (y/n)? > y
> [INFO] cfde_submit.client::check() Check PASSED, client is ready use 
> flows.
> [DEBUG] cfde_submit.client::start_deriva_flow() Startup: Validating input
> [DEBUG] cfde_submit.bdbag_utils::get_bag() Checking for a Git repository
> [DEBUG] cfde_submit.bdbag_utils::get_bag() Git repo found, collecting 
> metadata
> [DEBUG] cfde_submit.bdbag_utils::get_bag() Creating BDBag out of 
> directory '/home/jjyang/src/idg-cfde/data/targetpages684/submission'
> [DEBUG] cfde_submit.bdbag_utils::get_bag() Copying data to 
> '/home/jjyang/src/idg-cfde/data/targetpages684/submission_31cbcc7391bc7b73cf9c254e7297da3bc3f9c485' 
> before creating BDBag
> [DEBUG] cfde_submit.bdbag_utils::get_bag() BDBag created at 
> '/home/jjyang/src/idg-cfde/data/targetpages684/submission_31cbcc7391bc7b73cf9c254e7297da3bc3f9c485'
> [DEBUG] cfde_submit.bdbag_utils::get_bag() Archiving BDBag at 
> '/home/jjyang/src/idg-cfde/data/targetpages684/submission_31cbcc7391bc7b73cf9c254e7297da3bc3f9c485' 
> using 'zip'
> [DEBUG] cfde_submit.bdbag_utils::get_bag() BDBag archived to file 
> '/home/jjyang/src/idg-cfde/data/targetpages684/submission_31cbcc7391bc7b73cf9c254e7297da3bc3f9c485.zip'
> [DEBUG] cfde_submit.bdbag_utils::get_bag() Removing old directory 
> '/home/jjyang/src/idg-cfde/data/targetpages684/submission_31cbcc7391bc7b73cf9c254e7297da3bc3f9c485'
> [DEBUG] cfde_submit.validation::validate_user_submission() Validating 
> TableSchema in BDBag 
> '/home/jjyang/src/idg-cfde/data/targetpages684/submission_31cbcc7391bc7b73cf9c254e7297da3bc3f9c485.zip'
> TableSchema invalid due to the following errors:
>
>

**Reply by ACharbonneau** [reply]
Hey Jeremy,

It looks like there are two parallel issues here. The first, is that you need to fill 3 additional tables, and the second is that your file table needs some edits.

---
## Additional Tables

You can read all the details for a minimal submission in the example [here](https://docs.nih-cfde.org/en/latest/c2m2/draft-C2M2_specification/#a-minimal-c2m2-submission), but the TLDR is:

- the most minimal valid submission possible would be to submit: a filled [id_namespace.tsv](https://github.com/nih-cfde/published-documentation/wiki/TableInfo:-id_namespace.tsv), a filled [primary_dcc_contact.tsv](https://github.com/nih-cfde/published-documentation/wiki/TableInfo:-primary_dcc_contact.tsv) and a filled [project.tsv](https://github.com/nih-cfde/published-documentation/wiki/TableInfo:-project.tsv) table, plus the [JSON file](https://osf.io/e5tc2/) and header-only (so empty) versions of the remaining tables. This would essentially declare placeholder information: a namespace for your DCC, who you are, and what your project name is, without supplying any information about assets. So, it wouldn't be terribly useful, but it's the smallest possible amount of information to supply.

- the minimal valid submission that provides data, would be to supply those same 3 filled tables ([id_namespace.tsv](https://github.com/nih-cfde/published-documentation/wiki/TableInfo:-id_namespace.tsv), [primary_dcc_contact.tsv](https://github.com/nih-cfde/published-documentation/wiki/TableInfo:-primary_dcc_contact.tsv) and [project.tsv](https://github.com/nih-cfde/published-documentation/wiki/TableInfo:-project.tsv)) PLUS fill at least the required columns for one core entity table (so any of [file.tsv](https://github.com/nih-cfde/published-documentation/wiki/TableInfo:-file.tsv), [subject.tsv](https://github.com/nih-cfde/published-documentation/wiki/TableInfo:-subject.tsv) or [biosample.tsv](https://github.com/nih-cfde/published-documentation/wiki/TableInfo:-biosample.tsv)). This would not only declare the placeholder info, but also list one asset type.

- To improve your model after that, you would:
     - fill in some non-required columns for tables you are already using 
     and/or
     - fill one (or both) of the other core tables

- After that further improvements could be made by:
     - filling the empty tables which describe associations between the core table entities
     and/or
     - continuing to fill in non-required columns

Not all DCCs will necessarily use every table or every column in every table. Especially as we add complexity to the C2M2, we expect that some concepts just won't be useful for everyone, but that's the general process.

---

## File table edits

We do strict validation of the JSON schema when you do a submission. A little of that validation happens client side, and a lot more happens on our server after your package has been uploaded. The problem you're seeing is before the package uploads, so it's due to mis-matches between the requirements in the JSON schema and the data in your file.tsv.

For file.tsv the validator is expecting a file with exactly these column names, in exactly this order: 
```
id_namespace, local_id, project_id_namespace, project_local_id, persistent_id, creation_time, size_in_bytes, uncompressed_size_in_bytes, sha256, md5, filename, file_format, data_type, assay_type, mime_type
```
You can see a detailed description of those values [here](https://github.com/nih-cfde/published-documentation/wiki/TableInfo:-file.tsv) or for all the tables by clicking through the links in the [table summary page](https://github.com/nih-cfde/published-documentation/wiki/C2M2-Table-Summary). In the file.tsv table you uploaded, there are several columns missing, so it's not passing the validator.

When I run the validator locally, I'm also seeing a lot of schema type errors, so it's saying that your row values have the wrong format. I think that's just because it is reading them from the wrong columns, because the missing columns have shifted everything to the left, but if you continue to get schema errors after adding in the missing columns in, it would probably be worth double checking that the values match the specification.

**Reply by jeremyjyang** [reply]
That helps, thanks. I"ve added and reordered columns in file.tsv accordingly. Also checked the specification against our values. Still not working though, no details, just

> TableSchema invalid due to the following errors: 

Revised submission at 

https://www.dropbox.com/s/5s2d1llkmw1300x/idg_submission.zip?dl=0

Non-empty TSVs also in this spreadsheet 

https://docs.google.com/spreadsheets/d/1Mt9mtfNANIgnFvMIwH9az5QFAgcRxD-asn7TbBUzqhU/edit?usp=sharing

**Reply by ACharbonneau** [reply]
Hey Jeremy. 

I found schema errors in two tables:

> invalid: primary_dcc_contact.tsv
>   2  None   foreign-key-error  Row at position "2" violates the foreign key: not found in the lookup table
> 

>  invalid: project.tsv
> None      7  missing-label    There is a missing label in the header's field "description" at position "7"
> None      6  incorrect-label  Label "namedescription" in field name at position "6" does not match the field name in the schema
>    2      7  missing-cell     Row at position "2" has a missing cell in field "description" at position "7"
> 

For `primary_dcc_contact.tsv` the issue is that the project_local_id doesn't exist in your project table. The local_id for the project in your project table is targetpages, so it needs to be that in the primary_dcc_contact.tsv

For project.tsv there is no tab between the name and description columns, so it's treating them as one. 

I locally fixed both of these, and the resulting package submits fine to my test account, but client side validation only checks for schema errors. Server side additionally does value checking, and there I find additional errors:

These all seem to be about the values you have for id_namespacees in the file.tsv table. They don't match what is in id_namespace.tsv. If you are only using one id_namespace (which is what you currently have in id_namespace.tsv) then you need to use that as the value for id_namespace everywhere (so id_namespace and project_id_namespace). It's possible there are others, but it's getting stuck at the namespace issue, so these need to be fixed, so it can get farther into the check.

**Reply by jeremyjyang** [reply]
Thanks again. After correcting these mistakes, we're getting much further. Now the error is:


> ERROR:  insert or update on table "file" violates foreign key constraint "file_file_format_fkey". DETAIL:  Key (file_format)=(format:3464) is not present in table "file_format"

Since this format is JSON (https://www.ebi.ac.uk/ols/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fformat_3464) shouldn't that be present in the table already? Or do I need to include it?


> [INFO] cfde_submit.globus_http::upload() Upload successful to 'https://g-882990.aa98d.08cc.data.globus.org/CFDE/data/submission_8b1b1172c2ea67b5fc7af7ece63d1317c281be6b.zip': 200 b''
> [DEBUG] cfde_submit.client::start_deriva_flow() Flow input populated:
> {
>     "cfde_ep_id": "d4c89edc-a22c-4bc3-bfa2-bca5fd19b404",
>     "data_url": "https://g-882990.aa98d.08cc.data.globus.org/CFDE/data/submission_8b1b1172c2ea67b5fc7af7ece63d1317c281be6b.zip",
>     "dcc_id": "cfde_registry_dcc:idg",
>     "source_endpoint_id": false,
>     "test_sub": false
> }
> [DEBUG] cfde_submit.client::start_deriva_flow() Starting Flow - Submitting data
> [DEBUG] cfde_submit.client::start_deriva_flow() Flow started successfully.
> Started DERIVA ingest flow
> Your dataset has been submitted
> You can check the progress with: cfde-submit status
> 
> [DEBUG] cfde_submit.main::run() State saved to '/home/jjyang/.cfde_client.json'
> 
> Status of CFDE Submission (Flow ID afb6b637-c984-4daf-94ed-1d5922220c5a)
> This instance ID: 43251055-03bc-4d1b-8c61-54a41c8010c8
> 
> This Flow is still in progress.
> Current Flow Step: DerivaIngest
> 
> (cfde) lengua$ cfde-submit status
> 
> Status of CFDE Submission (Flow ID afb6b637-c984-4daf-94ed-1d5922220c5a)
> This instance ID: 43251055-03bc-4d1b-8c61-54a41c8010c8
> 
> This Flow has completed.
> Your CFDE submission () failed to ingest into DERIVA with this error:
> Error ingesting to DERIVA: Table "file" load error: 409 Conflict. The request conflicts with the state of the server. Input data violates model. ERROR:  insert or update on table "file" violates foreign key constraint "file_file_format_fkey". DETAIL:  Key (file_format)=(format:3464) is not present in table "file_format"

**Reply by ACharbonneau** [reply]
Not sure if Arthurs answers will give you a notification, so I'm making a notification appear :)

https://github.com/nih-cfde/published-documentation/discussions/184#discussioncomment-535849
https://github.com/nih-cfde/published-documentation/discussions/184#discussioncomment-535853

**Comment by abradyIGS** [comment]
hey there!

the term (3464) is in the EDAM reference as expected, but it's not in the `file_format.tsv` table included in your submission, according to that error. did you build that table (`file_format.tsv`) with the automated Python script (`build_term_tables.py` or something similar) before submitting? if not, that's the likely culprit.

(if you can link your current submission files somewhere i have read permissions, i can guess better, if need be.)

**Comment by abradyIGS** [comment]
(Less obtuse background detail: we ask as part of a valid submission that the CV term-tracking tables `file_format.tsv`, `anatomy.tsv`, etc., contain complete lists of all terms used elsewhere in the submission. If a term is used somewhere (in this case "`format:3464`" in the `file` table) and isn't _also_ found in the relevant CV table (in this case `file_format.tsv`), it generates the key-constraint error you reported above.)

**Reply by jeremyjyang** [reply]
Ok, with the needed file_format.tsv definition, we got in the end zone, and the victory dance commenced. 

Thanks.

Now on to the approval process. 

> (cfde) lengua$ cfde-submit status
> 
> Status of CFDE Submission (Flow ID afb6b637-c984-4daf-94ed-1d5922220c5a)
> This instance ID: d0a6ac9f-312c-4185-b835-df0a29f3c519
> 
> This Flow has completed.
> Submission Flow succeeded. Your submission ID is 3a2e3760-8e6b-11eb-8e7d-b37da3bc8f6e, and your submission can be viewed at this link: https://app.nih-cfde.org/chaise/recordset/#112/CFDE:file
